File: temp_peltier_server.py
#!/usr/bin/env python3
"""
Water temperature / Peltier monitoring server.
Separate from the pH server (that one is on :3000, this one is :3001).

Run with:   python3 temp_peltier_server.py
Requires:   pip3 install flask

The ESP32 POSTs JSON like {"temp": 24.87, "peltier": true} to /update.
Open http://<this-pi's-ip>:3001/ in a browser for a live status page.
"""

from flask import Flask, request, jsonify, render_template_string
from datetime import datetime
import threading

app = Flask(__name__)

PORT = 3001

lock = threading.Lock()
latest = {
    "temp": None,
    "peltier": None,
    "last_updated": None,
}

DASHBOARD_HTML = """
<!DOCTYPE html>
<html>
<head>
    <title>Water Temp / Peltier Monitor</title>
    <meta http-equiv="refresh" content="5">
    <style>
        body { font-family: sans-serif; background: #111; color: #eee; text-align: center; padding-top: 60px; }
        .temp { font-size: 4em; margin: 10px 0; }
        .peltier { font-size: 1.5em; padding: 8px 20px; border-radius: 8px; display: inline-block; }
        .on  { background: #2e7d32; }
        .off { background: #444; }
        .updated { color: #999; margin-top: 20px; font-size: 0.9em; }
    </style>
</head>
<body>
    <h1>Water Temperature Monitor</h1>
    <div class="temp">{{ temp if temp is not none else "--" }} &deg;C</div>
    <div class="peltier {{ 'on' if peltier else 'off' }}">
        Peltier: {{ "ON" if peltier else "OFF" }}
    </div>
    <div class="updated">Last updated: {{ last_updated or "no data yet" }}</div>
</body>
</html>
"""


@app.route("/update", methods=["POST"])
def update():
    data = request.get_json(silent=True)

    if not data or "temp" not in data:
        return jsonify({"error": "expected JSON body with a 'temp' field"}), 400

    with lock:
        latest["temp"] = data.get("temp")
        latest["peltier"] = bool(data.get("peltier", False))
        latest["last_updated"] = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        snapshot = dict(latest)

    print(f"[{snapshot['last_updated']}] temp={snapshot['temp']}  "
          f"peltier={'ON' if snapshot['peltier'] else 'OFF'}")

    return jsonify({"status": "ok"}), 200


@app.route("/status")
def status():
    with lock:
        return jsonify(latest)


@app.route("/")
def dashboard():
    with lock:
        data = dict(latest)
    return render_template_string(DASHBOARD_HTML, **data)


if __name__ == "__main__":
    # host="0.0.0.0" is required so the ESP32 can reach this over the LAN -
    # the default 127.0.0.1 only accepts connections from the Pi itself.
    app.run(host="0.0.0.0", port=PORT, debug=False)

File: temp_peltier_server.js
// Water temp / Peltier monitoring server - Node.js + Express + MariaDB
//
// Setup:
//   npm install
//   npm start
//
// Requires a MariaDB database + user already created (see setup guide).
// The 'readings' table is created automatically on first run if missing.

const express = require('express');
const mysql = require('mysql2/promise');

const app = express();
app.use(express.json());

const PORT = 3001;

// TODO: update these to match the database/user you created in MariaDB
const DB_CONFIG = {
    host: 'localhost',
    user: 'water_monitor',
    password: 'CHANGE_ME',
    database: 'water_monitor',
};

let pool;

async function initDb() {
    pool = mysql.createPool(DB_CONFIG);

    await pool.query(`
        CREATE TABLE IF NOT EXISTS readings (
            id INT AUTO_INCREMENT PRIMARY KEY,
            temp FLOAT NOT NULL,
            peltier BOOLEAN NOT NULL,
            created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
        )
    `);

    console.log('Connected to MariaDB, readings table ready');
}

// =============================
// ESP32 posts new readings here
// =============================
app.post('/update', async (req, res) => {
    const { temp, peltier } = req.body;

    if (typeof temp !== 'number') {
        return res.status(400).json({ error: "expected JSON body with a numeric 'temp' field" });
    }

    try {
        await pool.query(
            'INSERT INTO readings (temp, peltier) VALUES (?, ?)',
            [temp, !!peltier]
        );

        console.log(`[${new Date().toISOString()}] temp=${temp}  peltier=${peltier ? 'ON' : 'OFF'}`);

        res.json({ status: 'ok' });
    } catch (err) {
        console.error('DB insert failed:', err.message);
        res.status(500).json({ error: 'database error' });
    }
});

// =============================
// Latest reading as JSON
// =============================
app.get('/status', async (req, res) => {
    try {
        const [rows] = await pool.query(
            'SELECT temp, peltier, created_at FROM readings ORDER BY id DESC LIMIT 1'
        );
        res.json(rows[0] || { temp: null, peltier: null, created_at: null });
    } catch (err) {
        console.error('DB read failed:', err.message);
        res.status(500).json({ error: 'database error' });
    }
});

// =============================
// Last 100 readings as JSON
// =============================
app.get('/history', async (req, res) => {
    try {
        const [rows] = await pool.query(
            'SELECT temp, peltier, created_at FROM readings ORDER BY id DESC LIMIT 100'
        );
        res.json(rows);
    } catch (err) {
        console.error('DB read failed:', err.message);
        res.status(500).json({ error: 'database error' });
    }
});

// =============================
// Live dashboard
// =============================
app.get('/', async (req, res) => {
    let latest = { temp: null, peltier: null, created_at: null };

    try {
        const [rows] = await pool.query(
            'SELECT temp, peltier, created_at FROM readings ORDER BY id DESC LIMIT 1'
        );
        if (rows[0]) latest = rows[0];
    } catch (err) {
        console.error('DB read failed:', err.message);
    }

    res.send(`<!DOCTYPE html>
<html>
<head>
    <title>Water Temp / Peltier Monitor</title>
    <meta http-equiv="refresh" content="5">
    <style>
        body { font-family: sans-serif; background: #111; color: #eee; text-align: center; padding-top: 60px; }
        .temp { font-size: 4em; margin: 10px 0; }
        .peltier { font-size: 1.5em; padding: 8px 20px; border-radius: 8px; display: inline-block; }
        .on  { background: #2e7d32; }
        .off { background: #444; }
        .updated { color: #999; margin-top: 20px; font-size: 0.9em; }
    </style>
</head>
<body>
    <h1>Water Temperature Monitor</h1>
    <div class="temp">${latest.temp !== null ? latest.temp : '--'} &deg;C</div>
    <div class="peltier ${latest.peltier ? 'on' : 'off'}">Peltier: ${latest.peltier ? 'ON' : 'OFF'}</div>
    <div class="updated">Last updated: ${latest.created_at || 'no data yet'}</div>
</body>
</html>`);
});

initDb()
    .then(() => {
        // host 0.0.0.0 is required so the ESP32 can reach this over the LAN
        app.listen(PORT, '0.0.0.0', () => {
            console.log(`Server listening on port ${PORT}`);
        });
    })
    .catch((err) => {
        console.error('Failed to connect to database:', err.message);
        process.exit(1);
    });
